### DISCUSSION TITLE
关于如何实现一个 array function

### DISCUSSION THREAD

**Discussion by niebayes** [apache/doris#20276] [A - General / Q&A]
我需要实现 `array_contains_all` 函数，这个函数的签名为 `array_contains_all(Array<T> a, Array<T> b)`。如果数组 a 包含数组 b 中所有元素，则返回 1，否则返回 0。特别地，这个函数需要支持输入多维数组，且需要允许数组内元素为不同类型。

在尝试实现这个函数的过程中，我遇到了一些问题：

（1）关于 Block, Column, ColumnNullable, ColumnArray, ColumnVector，我的理解如下：
- Block：由多个 columns 组成，它的存在是为了可以批量处理数据。
- Column：内存中的一个数据块，本身并不提供类似 raw array 的语义。任何时候，当我们需要表示和存储一块数据时，我们使用 Column。
- ColumnNullable: 表示一个可以存储 NULL values 的数据块，其由 nested_column data 数据块，和 null_map  metadata 数据块组成。null_map 是一个 key-value map，key 为 chunk 的 index，value 表示该 chunk 是否为 NULL value。其中 chunk 为 nested_column data 数据块中的子数据块。任何时候，当我们需要表示一块可能为 NULL value 的数据块时，我们使用 ColumnNullable。
- ColumnArray：表示一个具有 raw array 语义的数据块。其内含 data 数据块，和 offsets metadata 数据块。offsets 由一系列 offsets 组成，每个 offset 指示 data 数据块中的子数据块的位置。data 数据块也使用 Column （的继承类）表示，也就是说 data 可以为 ColumnNullable。
- ColumnVector：表示一个具有 raw array 语义的数据块，同时使用一个 raw array（C++ 的 vector）来存储 data。

例如，当我们在 client 输入 `SELECT array_contains([1, NULL, 3], 1)` 时，Doris 会这样去表示所输入的 array：
- 使用 ColumnNullable 表示函数的第一个参数，因为输入数组可以为 NULL。这个 ColumnNullable instance 的 data 的类型为 ColumnArray，因为第一个参数为数组，不妨用 X 表示这个 data。这个 instance 的 null map 为空，因为第一个参数不为 NULL。
- X 的 data 为 ColumnNullable，因为输入数组可以包含 NULL values，不妨用 Y 表示这个 data。X 的 offsets 的 size 为 3，因为输入数组包含三个元素。
- Y 的 data 为 ColumnVector，vector 中的元素为输入数组中的元素（实际为指针）。Y 的 null map 为 `010`，表示左起第二个元素为 NULL value。

我的问题为：
- 我的理解是否正确？
- 如何用现有的这些结构表示一个多维数组？
- 目前前端 SQL 解析是否支持多维数组？
- 在 ColumnArray 类的注释中，有这么一句话：`the ColumnArray won't nest multi-layers. That means the nested type will be concrete data-type`，这是否表示 Doris 目前 array functions 的代码框架并不支持存储和表示多维数组？
- 关于 Column Array 的注释，其还有这么一句话：`In memory, it is represented as one column of a nested type, whose size is equal to the sum of the sizes of all arrays`， 这里是否应该是 `the sum of the sizes of all elements` ？还是说这是为了说明 ColumnArray 可以表示 an array of arrays？

（2）在执行 `execute_impl` 函数之前，block 是否已经被扩展了。在我 debug 时，我输入了两个数组参数，但是 block 的 size 为 4。这是为什么？我知道其中一个 column 是用来存储 result column，那么另一个呢？

（3）为什么很多 array functions 的实现中都有 `convert_to_full_column_if_const` 这个函数？我认为这个函数会导致 runtime overhead 增大。这里是否牺牲了 runtime overhead 来获取更高的 I/O 效率？P.S. 我在某个 PR 中看到了推荐使用 `unpack_if_const` 函数。

（4）arguments 参数表示什么？是一个 array of column index 吗？arguments 与 block 中的 columns 有何关系？

（5）debug 时很多 variable 被优化掉了，有办法取消这些优化吗？

（6）debug 时会遇到很多 pointers，有办法 dereference 这些 pointers，拿到它们指向的值吗？或者 Doris 有提供什么打印的方法吗？

（7）`get_return_type_impl` 是用来干啥的？为什么很多 array functions 的实现在其中检查输入参数的合法性？对于合法性的检查，为什么不在 `execute_impl` 函数中进行？

（8）对于 offsets 的 index，我们允许 `offsets[-1]`，`offsets[offsets.size()]` 的出现。为什么会这样设计？

（9）如果我要比较两个 data type 是否可以相互 cast，是不是应该使用 get_least_supertype 检查它们是否有公共祖先？

（10）通过学习其他 array functions 的 PR，我了解到需要修改 `doris_builtin_functions.py` 文件。我在其中添加了如下内容：
> [['array_contains_all'], 'BOOLEAN', ['ARRAY_BOOLEAN', 'ARRAY_BOOLEAN'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_TINYINT', 'ARRAY_TINYINT'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_SMALLINT', 'ARRAY_SMALLINT'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_INT', 'ARRAY_INT'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_BIGINT', 'ARRAY_BIGINT'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_LARGEINT', 'ARRAY_LARGEINT'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_DATETIME', 'ARRAY_DATETIME'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_DATE', 'ARRAY_DATE'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_DATETIMEV2', 'ARRAY_DATETIMEV2'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_DATEV2', 'ARRAY_DATEV2'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_FLOAT', 'ARRAY_FLOAT'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_DOUBLE', 'ARRAY_DOUBLE'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_DECIMALV2', 'ARRAY_DECIMALV2'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_DECIMAL32', 'ARRAY_DECIMAL32'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_DECIMAL64', 'ARRAY_DECIMAL64'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_DECIMAL128', 'ARRAY_DECIMAL128'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_VARCHAR', 'ARRAY_VARCHAR'], 'ALWAYS_NULLABLE'],
    [['array_contains_all'], 'BOOLEAN', ['ARRAY_STRING', 'ARRAY_STRING'], 'ALWAYS_NULLABLE'],

请问这样是否就能支持数组包含不同类型的元素？是否能支持多维数组？

**Comment by amorynan** [comment]
1.null_map 就是一个uint8 数组， 
2. 多维数组就是 Array(Array....)
